pink/barriers/body_spherical_barrier.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2022 Stéphane Caron
+
+"""Barrier based on the distance between two frames."""
+
+from typing import Tuple, Union
+
+import numpy as np
+
+from ..configuration import Configuration
+from .barrier import Barrier
+from .exceptions import NegativeMinimumDistance
+
+
+class BodySphericalBarrier(Barrier):
+    r"""A barrier based on the distance between two frames.
+
+    Defines a barrier function based on the Euclidean distance between two
+    specified frames. It allows for the specification of a minimum distance
+    threshold.
+
+    The barrier function is defined as:
+
+    .. math::
+
+        h(q) = \|p_1(q) -
+        p_2(q)\|^2 - d_{min}^2
+
+    where :math:`p_1(q)` and
+    :math:`p_2(q)` are the positions of the two
+    frames in the world coordinate system, and :math:`d_{min}`
+    is the minimum distance threshold.
+
+    Attributes:
+        frames: Tuple of two frame names.
+        d_min: Minimum distance threshold.
+    """
+
+    frames: Tuple[str, str]
+    d_min: float
+
+    def __init__(
+        self,
+        frames: Tuple[str, str],
+        d_min: float,
+        gain: Union[float, np.ndarray] = 1.0,
+        safe_displacement_gain: float = 3.0,
+    ):
+        """Initialize the BodySphericalBarrier.
+
+        Args:
+            frames: Tuple of two frame names.
+            d_min: Minimum distance threshold.
+            gain: Barrier gain. Defaults to 1.0.
+            safe_displacement_gain: gain for the safe backup displacement
+                cost term. Defaults to 3.0.
+        """
+        if d_min < 0.0:
+            raise NegativeMinimumDistance(
+                "The minimum distance threshold must be non-negative."
+            )
+        super().__init__(
+            dim=1,
+            gain=gain,
+            gain_function=lambda h: h / (1 + np.abs(h)),
+            safe_displacement_gain=safe_displacement_gain,
+        )
+        self.frames = frames
+        self.d_min = d_min
+
+    def compute_barrier(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the value of the barrier function.
+
+        The barrier function is computed based on the Euclidean distance
+        between the two specified frames. It considers the minimum distance
+        threshold.
+
+        The barrier function is given by:
+
+        .. math::
+
+            h(q) = \|p_1(q) -
+            p_2(q)\|^2 - d_{min}^2
+
+        where :math:`p_1(q)` and
+        :math:`p_2(q)` are the positions of the two
+        frames in the world coordinate system, and :math:`d_{min}`
+        is the minimum distance threshold.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Value of the barrier function
+                :math:`h(q)`.
+        """
+        pos1_world, pos2_world = self._get_frame_positions(configuration)
+        return np.array(
+            [np.linalg.norm(pos1_world - pos2_world) ** 2 - self.d_min**2]
```

Comment:
This is a minor suggestion but you can check that squaring a `linalg.norm` is slower than taking a dot product of a vector by itself:
```py
In [4]: %timeit np.linalg.norm(p1 - p2)**2
1.44 µs ± 16.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [5]: %timeit (p1-p2).dot(p1-p2)
845 ns ± 5.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```